changelog/deprecate_bad_module_name.dd

```diff
@@ -0,0 +1,47 @@
+Deprecate import module name mismatch
+
+Some instances of module name/import name mismatches were deprecated while others are now considered errors.
+
+Case 1: DEPRECATED: import a module name with a qualified name but the module being imported has no module declaration.
+---
+// main.d
+import foo.bar;
+
+// foo/bar.d
+// this file is empty, it has no module declaration
+---
+The above code will now print:
+
+$(CONSOLE Deprecated: module bar from file foo/bar.d should be imported with 'import bar;' )
+
+Case 2: DEPRECATED: import a module with a qualified name that partially matches the name of the module being imported.
+---
+// main.d
+import foo.bar;
+import foo.baz.buz;
+
+// foo/bar.d
+module bar;
+
+// foo/baz/buz.d
+module baz.buz;
+---
+The above code will now print:
+
+$(CONSOLE Deprecated: module bar from file foo/bar.d should be imported with 'import bar;'
```

Comment:
Alternatively: replace module declaration with `module foo.bar;`

We need to show advantages of this as well. For example show how the current rules lead to a breach in modularity.

My current mental model goes like this: `main.d` uses a dotted import to guide path selection, and that also confers a local name for the imported module. The fact that the module gave itself a non-dotted name is irrelevant to `main.d`. I see nothing wrong in the fact that the module name from the perspective of code inside `foo/bar.d` is different from the module name from the perspective of code inside `main.d`. In fact we have a language construct that messes exactly with that:

```d
import takezat = foo.bar;
```

We'd need a good argument that the above is a bad thing.

Reply:
Interesting.  That's not how I originally understood the module system to work, but your description matches the implementation that I found.

One disadvantage of the current mechanism is that this causes the compiler to load the same module multiple times.  Each unique path used to import the module will result in another instance of the module being instantiated.  I think there's more to say but I'll leave it at that for now.